legacy/reqcounter/reqcounter.go

```diff
@@ -19,76 +19,127 @@ package reqcounter
 import (
 	"fmt"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/sirupsen/logrus"
 )
 
 // RequestCounter records the number of HTTP requests to GCR.
+// TODO: @tylerferrara in the future, add a field 'persistent bool' to determine if
+// the counter should ever reset.
 type RequestCounter struct {
-	mutex    sync.Mutex // Lock to prevent race-conditions with concurrent processes.
-	requests uint64     // Number of HTTP requests since recording started.
-	since    time.Time  // When the current counter began recording requests.
+	Mutex    sync.Mutex    // Lock to prevent race-conditions with concurrent processes.
+	Requests uint64        // Number of HTTP requests since recording started.
+	Since    time.Time     // When the current counter began recording requests.
+	Interval time.Duration // The duration of time between each log.
 }
 
-// log records the number of HTTP requests found and resets the counter.
-func (rc *RequestCounter) log() {
+// increment adds 1 to the request counter, signifying another call to GCR.
+func (rc *RequestCounter) Increment() {
+	rc.Mutex.Lock()
+	atomic.AddUint64(&rc.Requests, 1)
```

Comment:
https://stackoverflow.com/questions/47445344/is-there-a-difference-in-go-between-a-counter-using-atomic-operations-and-one-us

Using both the lock and the atomic addition seems redundant. See the post above for more info.

Reply:
Taking a second look, this does not appear to be a use case for atomic operations. This is because all manipulations (reads or writes) of RequestCounter must first hold the object's mutex.

We can't use atomic operations here because of the Flush function. It attempts to:
- Read requests
- Set requests = 0
- Set since = current time

The Flush function must do all of these things at once, ensuring no function can interject and (read or write) the object during execution. That is why holding a single mutex is the way to go. We should not conduct all of these operations atomically, because they MUST be grouped together.